### PR TITLE
Unpin changes

### PR DESCRIPTION
**Pull request by cramertj**
r? @aturon 

cc @withoutboats, @RalfJung, @pythonesque, #49150

### CODE REVIEW

**Comment by aturon** [comment]
lgtm

**Comment by aturon** [comment]
@bors: r+

(There should be plenty of time for further comments before this actually gets to the front of the queue...)

**Comment by withoutboats** [comment]
:+1:

File: src/libcore/option.rs [hunk 1]
```diff
@@ -269,6 +270,15 @@ impl<T> Option<T> {
         }
     }
 
+    /// Converts from `Option<T>` to `Option<PinMut<'_, T>>`
```
Comment (RalfJung) [inline src/libcore/option.rs]: I think this should say "from `PinMut<'_, Option<T>` to".

**Comment by RalfJung** [comment]
> Add Option::as_pin_mut

This fixes the pinned invariant for `Option` to be structural, i.e., a pinned option has its content pinned. I feel this should be written down explicitly somewhere, maybe next to the `enum Option`? One consequence of this is that adding `impl<T> Unpin for Option<T>` would lead to unsoundness.

File: src/libcore/option.rs (hunk 1 above)
Comment (kennytm) [reply src/libcore/option.rs]: If this is the case, we should better change the docs of `as_ref()` and `as_mut()` as well.

**Comment by RalfJung** [comment]
> i'm also interested in doing the same for other structures in std, such as AssertUnwindSafe, Result, and Box

Why `Box`? I expected `Box` to be unconditionally `Unpin`, like other smart pointers. `Result` should be fine though and other non-pointer-wrappers.

**Comment by cramertj** [comment]
@RalfJung Why would `Box` and other smart pointers be unconditionally `Unpin`? Does it not seem useful to project in those cases? I had thought that `Arc` would even be unconditionally unpin and give access to `Pin` of the innards, although perhaps we'd want to make that into a wrapper type so that we could add `ArcBox` someday.

**Comment by cramertj** [comment]
@RalfJung Yes, that intuition makes sense. I had a different intuition, though-- I was thinking of `Box` as a sort of single-field wrapper type like `AssertUnwindSafe` that just happened to store its inner field in a separate allocation. Neither interpretation is more clearly correct to me than the other, and I don't see strong utility in being able to project into a box (just use `PInBox` if you don't need to move out) nor making `Box` unconditionally `Unpin` (what does being able to move out of a `PinMut<Box<NotUnpinType>>` get you?).

I lean slightly towards my original intuition (single-field wrapper with out-of-line storage) because it allows more broad trait impls and makes more code that *expects* to be able to project `PinMut` correct.

**Pull request merged**